src/lib/models/ConfigModel.ts

```diff
@@ -1,172 +1,194 @@
-  retire-js:
-
-    # Whether to turn off the 'retire-js' engine so that it is not included when running Code Analyzer commands
-    disable_engine: false`;
+function getDefaultRule(defaultState: ConfigState, engineName: string, ruleName: string): Rule|null {
+	try {
+		return defaultState.rules.getRule(engineName, ruleName);
+	} catch (e) {
+		return null;
+	}
+}
+
+function toYamlRule(userRule: Rule, defaultRule: Rule|null, styled: boolean): string {
+	const ruleName: string = userRule.getName();
+	const userSeverity: SeverityLevel = userRule.getSeverityLevel();
+	const userTags: string[] = userRule.getTags();
+
+	let severityYaml = yaml.dump({severity: userSeverity});
+	let tagsYaml = yaml.dump({tags: userTags});
+
+	if (defaultRule != null) {
+		const defaultSeverity: SeverityLevel = defaultRule.getSeverityLevel();
+		const defaultTagsJson: string = JSON.stringify(defaultRule.getTags());
+
+		if (userSeverity !== defaultSeverity) {
+			const comment = getMessage(BundleName.ConfigModel, 'template.modified-from', [defaultSeverity]);
+			severityYaml = severityYaml.replace('\n', ` ${toYamlComment(comment, styled)}\n`);
+		}
+
+		if (JSON.stringify(userTags) !== defaultTagsJson) {
+			const comment = getMessage(BundleName.ConfigModel, 'template.modified-from', [defaultTagsJson]);
+			// The YAML spec requires a trailing newline, so we know that there's at least one newline somewhere in the
+			// string. If we inject a comment before the first newline we encounter, then it will look clean.
+			tagsYaml = tagsYaml.replace('\n', ` ${toYamlComment(comment, styled)}\n`);
+		}
+	}
+	return `"${ruleName}":\n${indent(severityYaml, 2)}${indent(tagsYaml, 2)}`;
+}
+
+function toYamlEngines(userState: ConfigState, styled: boolean): string {
+	if (userState.rules.getCount() === 0) {
+		const comment = getMessage(BundleName.ConfigModel, 'template.yaml.no-engines-selected');
+		return `engines: {} ${toYamlComment(comment, styled)}`;
+	}
+
+	let results: string = 'engines:\n'
+
+	for (const engineName of userState.core.getEngineNames()) {
+		if (!userState.rules.getEngineNames().includes(engineName)) {
+			continue;
+		}
+		const engineConfigDescriptor = userState.core.getEngineConfigDescription(engineName);
+		const engineConfig = userState.core.getEngineConfig(engineName);
+
+		results += `\n${toYamlComment(getMessage(BundleName.ConfigModel, 'template.yaml.section-wrapper'), styled, 2)}\n`
+		// Engines are guaranteed to have an overview, even if it's just generic text.
+		results += `${toYamlComment(engineConfigDescriptor.overview!, styled, 2)}\n`;
+		results += `${toYamlComment(getMessage(BundleName.ConfigModel, 'template.yaml.section-wrapper'), styled, 2)}\n`
+
+		results += indent(`${engineName}:`, 2) + '\n';
+		// By fiat, the field description will always include, at minimum, an entry for "disable_field", so we can
+		// assume that the object is not undefined.
+		for (const configField of Object.keys(engineConfigDescriptor.fieldDescriptions!)) {
```

Comment:
Consider using `for (const [field, description] of Object.entries(engineConfigDescription.fieldDescriptions))` so that you don't need line 186.